diff --git a/HEAAN/lib/src/subdir.mk b/HEAAN/lib/src/subdir.mk
index da56a12..ac4eade 100644
--- a/HEAAN/lib/src/subdir.mk
+++ b/HEAAN/lib/src/subdir.mk
@@ -56,7 +56,7 @@ CPP_DEPS += \
 src/%.o: ../src/%.cpp
 	@echo 'Building file: $<'
 	@echo 'Invoking: GCC C++ Compiler'
-	g++ -I/usr/local/include -O3 -c -std=c++11 -pthread -MMD -MP -MF"$(@:%.o=%.d)" -MT"$(@)" -o "$@" "$<"
+	g++ -I/usr/local/include -g -c -std=c++11 -pthread -MMD -MP -MF"$(@:%.o=%.d)" -MT"$(@)" -o "$@" "$<"
 	@echo 'Finished building: $<'
 	@echo ' '
 
diff --git a/HEAAN/src/Ciphertext.cpp b/HEAAN/src/Ciphertext.cpp
index 14a1a32..d95f70a 100644
--- a/HEAAN/src/Ciphertext.cpp
+++ b/HEAAN/src/Ciphertext.cpp
@@ -24,18 +24,38 @@ Ciphertext::Ciphertext(const Ciphertext& o) : logp(o.logp), logq(o.logq), n(o.n)
 	}
 }
 
-void Ciphertext::copyParams(Ciphertext& o) {
+void Ciphertext::copyParams(const Ciphertext& o) {
 	logp = o.logp;
 	logq = o.logq;
 	n = o.n;
 }
 
-void Ciphertext::copy(Ciphertext& o) {
+void Ciphertext::copy(const Ciphertext& o) {
 	copyParams(o);
-	for (long i = 0; i < N; ++i) {
-		ax[i] = o.ax[i];
-		bx[i] = o.bx[i];
-	}
+    if(o.cx) {
+        addCX();
+        for (long i = 0; i < N; ++i) {
+            ax[i] = o.ax[i];
+            bx[i] = o.bx[i];
+            cx[i] = o.cx[i];
+        }
+    }else{
+        dropCX();
+        for (long i = 0; i < N; ++i) {
+            ax[i] = o.ax[i];
+            bx[i] = o.bx[i];
+        }
+    }
+}
+
+void Ciphertext::addCX() {
+    if(!cx)
+        cx = new ZZ[N];
+}
+
+void Ciphertext::dropCX() {
+    delete[] cx;
+    cx = nullptr;
 }
 
 void Ciphertext::free() {
@@ -43,11 +63,20 @@ void Ciphertext::free() {
 		clear(ax[i]);
 		clear(bx[i]);
 	}
+    if(cx){
+        for(long i = 0; i < N; i++)
+            clear(cx[i]);
+    }
+}
+
+bool Ciphertext::relined() const{
+    return cx == nullptr;
 }
 
 Ciphertext::~Ciphertext() {
 	delete[] ax;
 	delete[] bx;
+    delete[] cx;
 }
 
 }  // namespace heaan
diff --git a/HEAAN/src/Ciphertext.h b/HEAAN/src/Ciphertext.h
index bd7fc78..ab35b18 100644
--- a/HEAAN/src/Ciphertext.h
+++ b/HEAAN/src/Ciphertext.h
@@ -20,6 +20,7 @@ public:
 
 	NTL::ZZ* ax = new NTL::ZZ[N];
 	NTL::ZZ* bx = new NTL::ZZ[N];
+    NTL::ZZ* cx = nullptr; // after ctxt multiplication
 
 	long logp;
 	long logq;
@@ -30,12 +31,18 @@ public:
 
 	Ciphertext(const Ciphertext& o);
 
-	void copyParams(Ciphertext& o);
+	void copyParams(const Ciphertext& o);
 
-	void copy(Ciphertext& o);
+	void copy(const Ciphertext& o);
+
+    void addCX();
+
+    void dropCX();
 
 	void free();
 
+    bool relined() const;
+
 	virtual ~Ciphertext();
 
 };
diff --git a/HEAAN/src/EvaluatorUtils.cpp b/HEAAN/src/EvaluatorUtils.cpp
index 4422404..0a7bab7 100644
--- a/HEAAN/src/EvaluatorUtils.cpp
+++ b/HEAAN/src/EvaluatorUtils.cpp
@@ -91,13 +91,13 @@ ZZ EvaluatorUtils::scaleUpToZZ(const RR& x, const long logp) {
 //----------------------------------------------------------------------------------
 
 
-void EvaluatorUtils::leftRotateAndEqual(complex<double>* vals, const long n, const long r) {
+void EvaluatorUtils::leftRotateAndEqual(complex<RR>* vals, const long n, const long r) {
 	long rem = r % n;
 	if(rem != 0) {
 		long divisor = GCD(rem, n);
 		long steps = n / divisor;
 		for (long i = 0; i < divisor; ++i) {
-			complex<double> tmp = vals[i];
+			complex<RR> tmp = vals[i];
 			long idx = i;
 			for (long j = 0; j < steps - 1; ++j) {
 				vals[idx] = vals[(idx + rem) % n];
@@ -108,7 +108,24 @@ void EvaluatorUtils::leftRotateAndEqual(complex<double>* vals, const long n, con
 	}
 }
 
-void EvaluatorUtils::rightRotateAndEqual(complex<double>* vals, const long n, const long r) {
+void EvaluatorUtils::leftRotateAndEqual(complex<double>* vals, const long n, const long r) {
+    long rem = r % n;
+    if(rem != 0) {
+        long divisor = GCD(rem, n);
+        long steps = n / divisor;
+        for (long i = 0; i < divisor; ++i) {
+            complex<double> tmp = vals[i];
+            long idx = i;
+            for (long j = 0; j < steps - 1; ++j) {
+                vals[idx] = vals[(idx + rem) % n];
+                idx = (idx + rem) % n;
+            }
+            vals[idx] = tmp;
+        }
+    }
+}
+
+void EvaluatorUtils::rightRotateAndEqual(complex<RR>* vals, const long n, const long r) {
 	long rem = r % n;
 	rem = (n - rem) % n;
 	leftRotateAndEqual(vals, n, rem);
diff --git a/HEAAN/src/EvaluatorUtils.h b/HEAAN/src/EvaluatorUtils.h
index 9e2d518..c841f47 100644
--- a/HEAAN/src/EvaluatorUtils.h
+++ b/HEAAN/src/EvaluatorUtils.h
@@ -40,7 +40,9 @@ public:
 	//   DOUBLE & RR <-> ZZ
 	//----------------------------------------------------------------------------------
 
-	static double scaleDownToReal(const NTL::ZZ& x, const long logp);
+//	static double scaleDownToReal(const NTL::ZZ& x, const long logp);
+
+    static double scaleDownToReal(const NTL::ZZ& x, const long logp);
 
 	static NTL::ZZ scaleUpToZZ(const double x, const long logp);
 
@@ -51,10 +53,10 @@ public:
 	//   ROTATIONS
 	//----------------------------------------------------------------------------------
 
+    static void leftRotateAndEqual(std::complex<double>* vals, const long n, const long r);
+	static void leftRotateAndEqual(std::complex<NTL::RR>* vals, const long n, const long r);
 
-	static void leftRotateAndEqual(std::complex<double>* vals, const long n, const long r);
-
-	static void rightRotateAndEqual(std::complex<double>* vals, const long n, const long r);
+	static void rightRotateAndEqual(std::complex<NTL::RR>* vals, const long n, const long r);
 
 };
 
diff --git a/HEAAN/src/Params.h b/HEAAN/src/Params.h
index 3924cff..af75c0c 100644
--- a/HEAAN/src/Params.h
+++ b/HEAAN/src/Params.h
@@ -10,10 +10,13 @@
 
 #include <NTL/ZZ.h>
 
+// alternative is NTL::RR
+#define BOOT_PREC_TYPE double
+
 namespace heaan {
 
 static const long logN = 16;
-static const long logQ = 800; // 128-bit security
+static const long logQ = 1250; // 128-bit security
 
 static const double sigma = 3.2;
 static const long h = 64;
diff --git a/HEAAN/src/Ring.cpp b/HEAAN/src/Ring.cpp
index 515ba7e..6b2436d 100644
--- a/HEAAN/src/Ring.cpp
+++ b/HEAAN/src/Ring.cpp
@@ -23,6 +23,10 @@ using namespace NTL;
 
 namespace heaan {
 
+std::complex<double> Ring::cmplxRR2double(const std::complex<NTL::RR> &val) {
+    return std::complex<double>(to_double(val.real()), to_double(val.imag()));
+}
+
 Ring::Ring() {
 
 	qpows = new ZZ[logQQ + 1];
@@ -39,9 +43,9 @@ Ring::Ring() {
 		fivePows %= M;
 	}
 
-	ksiPows = new complex<double>[M + 1];
+	ksiPows = new complex<BOOT_PREC_TYPE>[M + 1];
 	for (long j = 0; j < M; ++j) {
-		double angle = 2.0 * M_PI * j / M;
+		BOOT_PREC_TYPE angle = 2.0 * conv<BOOT_PREC_TYPE>(Pi) * j / M;
 		ksiPows[j].real(cos(angle));
 		ksiPows[j].imag(sin(angle));
 	}
@@ -62,6 +66,19 @@ void Ring::arrayBitReverse(complex<double>* vals, long n) {
 	}
 }
 
+void Ring::arrayBitReverse(complex<RR>* vals, long n) {
+    for (long i = 1, j = 0; i < n; ++i) {
+        long bit = n >> 1;
+        for (; j >= bit; bit >>= 1) {
+            j -= bit;
+        }
+        j += bit;
+        if (i < j) {
+            swap(vals[i], vals[j]);
+        }
+    }
+}
+
 void Ring::EMB(complex<double>* vals, long n) {
 	arrayBitReverse(vals, n);
 	for (long len = 2; len <= n; len <<= 1) {
@@ -73,7 +90,7 @@ void Ring::EMB(complex<double>* vals, long n) {
 				long idx = ((rotGroup[j] % lenq)) * gap;
 				complex<double> u = vals[i + j];
 				complex<double> v = vals[i + j + lenh];
-				v *= ksiPows[idx];
+				v *= cmplxRR2double(ksiPows[idx]);
 				vals[i + j] = u + v;
 				vals[i + j + lenh] = u - v;
 			}
@@ -91,7 +108,7 @@ void Ring::EMBInvLazy(complex<double>* vals, long n) {
 				long idx = (lenq - (rotGroup[j] % lenq)) * gap;
 				complex<double> u = vals[i + j] + vals[i + j + lenh];
 				complex<double> v = vals[i + j] - vals[i + j + lenh];
-				v *= ksiPows[idx];
+				v *= cmplxRR2double(ksiPows[idx]);
 				vals[i + j] = u;
 				vals[i + j + lenh] = v;
 			}
@@ -100,6 +117,25 @@ void Ring::EMBInvLazy(complex<double>* vals, long n) {
 	arrayBitReverse(vals, n);
 }
 
+void Ring::EMBInvLazy(complex<RR>* vals, long n) {
+    for (long len = n; len >= 1; len >>= 1) {
+        for (long i = 0; i < n; i += len) {
+            long lenh = len >> 1;
+            long lenq = len << 2;
+            long gap = M / lenq;
+            for (long j = 0; j < lenh; ++j) {
+                long idx = (lenq - (rotGroup[j] % lenq)) * gap;
+                complex<RR> u = vals[i + j] + vals[i + j + lenh];
+                complex<RR> v = vals[i + j] - vals[i + j + lenh];
+                v *= ksiPows[idx];
+                vals[i + j] = u;
+                vals[i + j + lenh] = v;
+            }
+        }
+    }
+    arrayBitReverse(vals, n);
+}
+
 void Ring::EMBInv(complex<double>* vals, long n) {
 	EMBInvLazy(vals, n);
 	for (long i = 0; i < n; ++i) {
@@ -107,8 +143,15 @@ void Ring::EMBInv(complex<double>* vals, long n) {
 	}
 }
 
+void Ring::EMBInv(complex<RR>* vals, long n) {
+    EMBInvLazy(vals, n);
+    for (long i = 0; i < n; ++i) {
+        vals[i] /= RR(n);
+    }
+}
+
 void Ring::encode(ZZ* mx, double* vals, long slots, long logp) {
-	complex<double>* uvals = new complex<double>[slots];
+	auto* uvals = new complex<double>[slots];
 	long i, jdx, idx;
 	for (i = 0; i < slots; ++i) {
 		uvals[i].real(vals[i]);
@@ -126,7 +169,7 @@ void Ring::encode(ZZ* mx, double* vals, long slots, long logp) {
 }
 
 void Ring::encode(ZZ* mx, complex<double>* vals, long slots, long logp) {
-	complex<double>* uvals = new complex<double> [slots];
+	auto* uvals = new complex<RR> [slots];
 	long i, jdx, idx;
 	copy(vals, vals + slots, uvals);
 	long gap = Nh / slots;
@@ -155,6 +198,12 @@ void Ring::decode(ZZ* mx, complex<double>* vals, long slots, long logp, long log
 }
 
 void Ring::addBootContext(long logSlots, long logp) {
+    /*
+     * functions called:
+     *  EvaluatorUtils::rightRotateAndEqual
+     *  EMBInv -> EMBInvLazy -> arrayBitReverse
+     *  EvaluatorUtils::scaleUpToZZ
+     * */
 	if (bootContextMap.find(logSlots) == bootContextMap.end()) {
 		long slots = 1 << logSlots;
 		long dslots = slots << 1;
@@ -177,9 +226,9 @@ void Ring::addBootContext(long logSlots, long logp) {
 		long bnd2;
 
 		ZZ* pvec = new ZZ[N];
-		complex<double>* pvals = new complex<double> [dslots];
+		auto* pvals = new complex<RR> [dslots];
 
-		double c = 0.25 / M_PI;
+		RR c = 0.25 / Pi;
 
 		if (logSlots < logNh) {
 			long dgap = gap >> 1;
@@ -214,9 +263,9 @@ void Ring::addBootContext(long logSlots, long logp) {
 			}
 
 			for (i = 0; i < slots; ++i) {
-				pvals[i] = 0.0;
-				pvals[i + slots].real(0);
-				pvals[i + slots].imag(-c);
+				pvals[i] = RR(0.0);
+				pvals[i + slots].real(RR(0));
+				pvals[i + slots].imag(RR(-c));
 			}
 			EMBInv(pvals, dslots);
 			for (i = 0, jdx = Nh, idx = 0; i < dslots; ++i, jdx += dgap, idx += dgap) {
@@ -232,8 +281,8 @@ void Ring::addBootContext(long logSlots, long logp) {
 			}
 
 			for (i = 0; i < slots; ++i) {
-				pvals[i] = c;
-				pvals[i + slots] = 0;
+				pvals[i] = RR(c);
+				pvals[i + slots] = RR(0);
 			}
 
 			EMBInv(pvals, dslots);
diff --git a/HEAAN/src/Ring.h b/HEAAN/src/Ring.h
index b2aa285..d74eaff 100644
--- a/HEAAN/src/Ring.h
+++ b/HEAAN/src/Ring.h
@@ -25,12 +25,13 @@ public:
 
 	NTL::ZZ* qpows;
 	long* rotGroup;
-	std::complex<double>* ksiPows;
+	std::complex<BOOT_PREC_TYPE>* ksiPows;
 	std::map<long, BootContext*> bootContextMap;
 	RingMultiplier multiplier;
 
 	Ring();
 
+    static std::complex<double> cmplxRR2double(const std::complex<NTL::RR>& val);
 
 	//----------------------------------------------------------------------------------
 	//   Encode & Decode
@@ -38,12 +39,16 @@ public:
 
 
 	void arrayBitReverse(std::complex<double>* vals, long size);
+    void arrayBitReverse(std::complex<NTL::RR>* vals, long size);
 
 	void EMB(std::complex<double>* vals, long size);
 
 	void EMBInvLazy(std::complex<double>* vals, long size);
+    void EMBInvLazy(std::complex<NTL::RR>* vals, long size);
+
 
 	void EMBInv(std::complex<double>* vals, long size);
+    void EMBInv(std::complex<NTL::RR>* val, long size);
 
 	void encode(NTL::ZZ* mx, double* vals, long slots, long logp);
 
diff --git a/HEAAN/src/Scheme.cpp b/HEAAN/src/Scheme.cpp
index 0b1fa9a..77392a6 100644
--- a/HEAAN/src/Scheme.cpp
+++ b/HEAAN/src/Scheme.cpp
@@ -208,7 +208,7 @@ void Scheme::encode(Plaintext& plain, complex<double>* vals, long n, long logp,
 }
 
 complex<double>* Scheme::decode(Plaintext& plain) {
-	complex<double>* res = new complex<double>[plain.n];
+	auto* res = new complex<double>[plain.n];
 	ring.decode(plain.mx, res, plain.n, plain.logp, plain.logq);
 	return res;
 }
@@ -380,7 +380,7 @@ complex<double> Scheme::decryptSingle(SecretKey& secretKey, Ciphertext& cipher)
 
 //-----------------------------------------
 
-void Scheme::negate(Ciphertext& res, Ciphertext& cipher) {
+void Scheme::negate(Ciphertext& res, const Ciphertext &cipher) {
 	res.copyParams(cipher);
 	ring.negate(res.ax, cipher.ax);
 	ring.negate(res.bx, cipher.bx);
@@ -391,36 +391,50 @@ void Scheme::negateAndEqual(Ciphertext& cipher) {
 	ring.negateAndEqual(cipher.bx);
 }
 
-void Scheme::add(Ciphertext& res, Ciphertext& cipher1, Ciphertext& cipher2) {
+void Scheme::add(Ciphertext& res, const Ciphertext &cipher1, const Ciphertext &cipher2) {
 	ZZ q = ring.qpows[cipher1.logq];
 	res.copyParams(cipher1);
 	ring.add(res.ax, cipher1.ax, cipher2.ax, q);
 	ring.add(res.bx, cipher1.bx, cipher2.bx, q);
-}
-
-void Scheme::addAndEqual(Ciphertext& cipher1, Ciphertext& cipher2) {
+    if(cipher2.cx || cipher1.cx){
+        res.addCX();
+        ZZ* src1 = cipher1.cx ? cipher1.cx : res.cx;
+        ZZ* src2 = cipher2.cx ? cipher2.cx : res.cx;
+        ring.add(res.cx, src1, src2, q);
+    }
+    else
+        res.dropCX();
+}
+
+void Scheme::addAndEqual(Ciphertext& cipher1, const Ciphertext &cipher2) {
+    if(cipher1.logp != cipher2.logp || cipher1.logq != cipher2.logq)
+        throw std::invalid_argument("params mismatch");
 	ZZ q = ring.qpows[cipher1.logq];
 	ring.addAndEqual(cipher1.ax, cipher2.ax, q);
 	ring.addAndEqual(cipher1.bx, cipher2.bx, q);
+    if(cipher2.cx){
+        cipher1.addCX();
+        ring.addAndEqual(cipher1.cx, cipher2.cx, q);
+    }
 }
 
 //-----------------------------------------
 
-void Scheme::addConst(Ciphertext& res, Ciphertext& cipher, double cnst, long logp) {
+void Scheme::addConst(Ciphertext& res, const Ciphertext &cipher, double cnst, long logp) {
 	ZZ q = ring.qpows[cipher.logq];
 	ZZ cnstZZ = logp < 0 ? EvaluatorUtils::scaleUpToZZ(cnst, cipher.logp) : EvaluatorUtils::scaleUpToZZ(cnst, logp);
 	res.copy(cipher);
 	AddMod(res.bx[0], res.bx[0], cnstZZ, q);
 }
 
-void Scheme::addConst(Ciphertext& res, Ciphertext& cipher, RR& cnst, long logp) {
+void Scheme::addConst(Ciphertext& res, const Ciphertext &cipher, const RR &cnst, long logp) {
 	ZZ q = ring.qpows[cipher.logq];
 	ZZ cnstZZ = logp < 0 ? EvaluatorUtils::scaleUpToZZ(cnst, cipher.logp) : EvaluatorUtils::scaleUpToZZ(cnst, logp);
 	res.copy(cipher);
 	AddMod(res.bx[0], res.bx[0], cnstZZ, q);
 }
 
-void Scheme::addConst(Ciphertext& res, Ciphertext& cipher, complex<double> cnst, long logp) {
+void Scheme::addConst(Ciphertext& res, const Ciphertext &cipher, complex<double> cnst, long logp) {
 	ZZ q = ring.qpows[cipher.logq];
 	ZZ cnstZZ = logp < 0 ? EvaluatorUtils::scaleUpToZZ(cnst.real(), cipher.logp) : EvaluatorUtils::scaleUpToZZ(cnst.real(), logp);
 	res.copy(cipher);
@@ -433,7 +447,9 @@ void Scheme::addConstAndEqual(Ciphertext& cipher, double cnst, long logp) {
 	AddMod(cipher.bx[0], cipher.bx[0], cnstZZ, q);
 }
 
-void Scheme::addConstAndEqual(Ciphertext& cipher, RR& cnst, long logp) {
+void Scheme::addConstAndEqual(Ciphertext& cipher, const RR &cnst, long logp) {
+    if(cipher.logp != logp)
+        throw std::invalid_argument("params mismatch");
 	ZZ q = ring.qpows[cipher.logq];
 	ZZ cnstZZ = logp < 0 ? EvaluatorUtils::scaleUpToZZ(cnst, cipher.logp) : EvaluatorUtils::scaleUpToZZ(cnst, logp);
 	AddMod(cipher.bx[0], cipher.bx[0], cnstZZ, q);
@@ -449,33 +465,33 @@ void Scheme::addConstAndEqual(Ciphertext& cipher, complex<double> cnst, long log
 
 //-----------------------------------------
 
-void Scheme::sub(Ciphertext& res, Ciphertext& cipher1, Ciphertext& cipher2) {
+void Scheme::sub(Ciphertext& res, const Ciphertext &cipher1, const Ciphertext &cipher2) {
 	ZZ q = ring.qpows[cipher1.logq];
 	res.copyParams(cipher1);
 	ring.sub(res.ax, cipher1.ax, cipher2.ax, q);
 	ring.sub(res.bx, cipher1.bx, cipher2.bx, q);
 }
 
-void Scheme::subAndEqual(Ciphertext& cipher1, Ciphertext& cipher2) {
+void Scheme::subAndEqual(Ciphertext& cipher1, const Ciphertext &cipher2) {
 	ZZ q = ring.qpows[cipher1.logq];
 	ring.subAndEqual(cipher1.ax, cipher2.ax, q);
 	ring.subAndEqual(cipher1.bx, cipher2.bx, q);
 }
 
-void Scheme::subAndEqual2(Ciphertext& cipher1, Ciphertext& cipher2) {
+void Scheme::subAndEqual2(Ciphertext& cipher1, const Ciphertext &cipher2) {
 	ZZ q = ring.qpows[cipher1.logq];
 	ring.subAndEqual2(cipher1.ax, cipher2.ax, q);
 	ring.subAndEqual2(cipher1.bx, cipher2.bx, q);
 }
 
-void Scheme::imult(Ciphertext& res, Ciphertext& cipher) {
+void Scheme::imult(Ciphertext& res, const Ciphertext &cipher) {
 	ZZ q = ring.qpows[cipher.logq];
 	res.copyParams(cipher);
 	ring.multByMonomial(res.ax, cipher.ax, Nh);
 	ring.multByMonomial(res.bx, cipher.bx, Nh);
 }
 
-void Scheme::idiv(Ciphertext& res, Ciphertext& cipher) {
+void Scheme::idiv(Ciphertext& res, const Ciphertext &cipher) {
 	ZZ q = ring.qpows[cipher.logq];
 	res.copyParams(cipher);
 	ring.multByMonomial(res.ax, cipher.ax, 3 * Nh);
@@ -492,7 +508,7 @@ void Scheme::idivAndEqual(Ciphertext& cipher) {
 	ring.multByMonomialAndEqual(cipher.bx, 3 * Nh);
 }
 
-void Scheme::mult(Ciphertext& res, Ciphertext& cipher1, Ciphertext& cipher2) {
+void Scheme::mult(Ciphertext& res, const Ciphertext &cipher1, const Ciphertext &cipher2) {
 	res.copyParams(cipher1);
 	res.logp += cipher2.logp;
 
@@ -546,7 +562,7 @@ void Scheme::mult(Ciphertext& res, Ciphertext& cipher1, Ciphertext& cipher2) {
 	delete[] raa;
 }
 
-void Scheme::multAndEqual(Ciphertext& cipher1, Ciphertext& cipher2) {
+void Scheme::multAndEqual(Ciphertext& cipher1, const Ciphertext &cipher2) {
 
 	ZZ q = ring.qpows[cipher1.logq];
 	ZZ qQ = ring.qpows[cipher1.logq + logQ];
@@ -603,7 +619,7 @@ void Scheme::multAndEqual(Ciphertext& cipher1, Ciphertext& cipher2) {
 
 //-----------------------------------------
 
-void Scheme::square(Ciphertext& res, Ciphertext& cipher) {
+void Scheme::square(Ciphertext& res, const Ciphertext &cipher) {
 	res.copyParams(cipher);
 	res.logp += cipher.logp;
 	ZZ q = ring.qpows[cipher.logq];
@@ -698,7 +714,7 @@ void Scheme::squareAndEqual(Ciphertext& cipher) {
 
 //-----------------------------------------
 
-void Scheme::multByConst(Ciphertext& res, Ciphertext& cipher, double cnst, long logp) {
+void Scheme::multByConst(Ciphertext& res, const Ciphertext &cipher, double cnst, long logp) {
 	ZZ q = ring.qpows[cipher.logq];
 	ZZ cnstZZ = EvaluatorUtils::scaleUpToZZ(cnst, logp);
 	ring.multByConst(res.ax, cipher.ax, cnstZZ, q);
@@ -707,7 +723,22 @@ void Scheme::multByConst(Ciphertext& res, Ciphertext& cipher, double cnst, long
 	res.logp += logp;
 }
 
-void Scheme::multByConst(Ciphertext& res, Ciphertext& cipher, complex<double> cnst, long logp) {
+void Scheme::multByConst(Ciphertext &res, const Ciphertext &cipher, const NTL::RR &cnst, long logp) {
+    ZZ q = ring.qpows[cipher.logq];
+    ZZ cnstZZ = EvaluatorUtils::scaleUpToZZ(cnst, logp);
+    ring.multByConst(res.ax, cipher.ax, cnstZZ, q);
+    ring.multByConst(res.bx, cipher.bx, cnstZZ, q);
+    if(cipher.cx){
+        res.addCX();
+        ring.multByConst(res.cx, cipher.cx, cnstZZ, q);
+    }
+    else
+        res.dropCX();
+    res.copyParams(cipher);
+    res.logp += logp;
+}
+
+void Scheme::multByConst(Ciphertext& res, const Ciphertext &cipher, complex<double> cnst, long logp) {
 	res.copy(cipher);
 	multByConstAndEqual(res, cnst, logp);
 }
@@ -733,11 +764,13 @@ void Scheme::multByConstAndEqual(Ciphertext& cipher, double cnst, long logp) {
 	cipher.logp += logp;
 }
 
-void Scheme::multByConstAndEqual(Ciphertext& cipher, RR& cnst, long logp) {
+void Scheme::multByConstAndEqual(Ciphertext& cipher, const RR& cnst, long logp) {
 	ZZ q = ring.qpows[cipher.logq];
 	ZZ cnstZZ = EvaluatorUtils::scaleUpToZZ(cnst, logp);
 	ring.multByConstAndEqual(cipher.ax, cnstZZ, q);
 	ring.multByConstAndEqual(cipher.bx, cnstZZ, q);
+    if(cipher.cx)
+        ring.multByConstAndEqual(cipher.cx, cnstZZ, q);
 	cipher.logp += logp;
 }
 
@@ -852,15 +885,194 @@ void Scheme::divByPo2AndEqual(Ciphertext& cipher, long bits) {
 	cipher.logq -= bits;
 }
 
+// more accurate ctxt multiplication: postpone rescaling
+// bits per level = logp
+void Scheme::matchScalingFactor(Ciphertext &dst1, Ciphertext &dst2, const Ciphertext &src1, const Ciphertext &src2, long logp) {
+    // TODO: optimize: avoid creating and copying ctxts
+    dst1.copy(src1);
+    dst2.copy(src2);
+    // compare the capacity of the ctxts
+    Ciphertext* less_cap = &dst1, *more_cap = &dst2;
+    long diff_cap = (dst1.logq - dst1.logp) - (dst2.logq - dst2.logp);
+    if(diff_cap > 0){
+        less_cap = &dst2;
+        more_cap = &dst1;
+    } // TODO: is it necessary to choose the ctxt with smaller rounding error when diff_cap == 0?
+    // rescale them s.t.
+    // (1) all of less_cap's capacity is preserved
+    // (2) more_cap's scaling factor is `logp`
+    // (3) less_cap and more_cap have the same logq
+    if(more_cap->logp > logp)
+        reScaleByAndEqual(*more_cap, more_cap->logp - logp);
+    if(less_cap->logq > more_cap->logq){
+        reScaleByAndEqual(*less_cap, less_cap->logq - more_cap->logq);
+    }
+    else if(less_cap->logq < more_cap->logq){
+        modDownToAndEqual(*more_cap, less_cap->logq);
+    }
+}
+
+// ****************** raw mult *****************
+// inplace-safe
+// return (b1*b2, a1*b2 + a2*b1, a1*a2)
+void Scheme::multRaw(Ciphertext& res, const Ciphertext &cipher1, const Ciphertext &cipher2){
+    if(cipher1.cx || cipher2.cx)
+        throw std::invalid_argument("cannot multiply with 3-tuple ctxt");
+    if(cipher1.n != cipher2.n)
+        throw std::invalid_argument("params mismatch");
+    if(cipher1.logq != cipher2.logq){
+        fprintf(stderr, "warning: logq different\n");
+    }
+    res.n = cipher1.n;
+    res.logq = std::min(cipher1.logq, cipher2.logq);
+    res.logp = cipher1.logp +  cipher2.logp; // NOTE: for inplace safety
+    res.addCX();
+
+    ZZ q = ring.qpows[res.logq];
+
+    long np = ceil((2 + cipher1.logq + cipher2.logq + logN + 2)/(double)pbnd);
+
+    auto* ra1 = new uint64_t[np << logN];
+    auto* rb1 = new uint64_t[np << logN];
+    auto* ra2 = new uint64_t[np << logN];
+    auto* rb2 = new uint64_t[np << logN];
+
+    ring.CRT(ra1, cipher1.ax, np);
+    ring.CRT(rb1, cipher1.bx, np);
+    ring.CRT(ra2, cipher2.ax, np);
+    ring.CRT(rb2, cipher2.bx, np);
+
+    ring.multDNTT(res.cx, ra1, ra2, np, q);
+    ring.multDNTT(res.bx, rb1, rb2, np, q);
+    // karatsuba
+    ring.addNTTAndEqual(ra1, rb1, np);
+    ring.addNTTAndEqual(ra2, rb2, np);
+    ring.multDNTT(res.ax, ra1, ra2, np, q);
+    ring.subAndEqual(res.ax, res.cx, q);
+    ring.subAndEqual(res.ax, res.bx, q);
+
+    delete[] ra1;
+    delete[] ra2;
+    delete[] rb1;
+    delete[] rb2;
+}
+
+void Scheme::multRawInplace(Ciphertext& cipher1, const Ciphertext &cipher2){
+    multRaw(cipher1, cipher1, cipher2);
+}
+
+// inplace safe
+// return (b^2, 2ab, a^2)
+void Scheme::squareRaw(Ciphertext& res, const Ciphertext &cipher){
+    if(cipher.cx)
+        throw std::invalid_argument("cannot multiply with 3-tuple ctxt");
+    res.copyParams(cipher);
+    res.logp += cipher.logp;
+    res.addCX();
+
+    ZZ q = ring.qpows[cipher.logq];
+
+    long np = ceil((2 * cipher.logq + logN + 2)/(double)pbnd);
+
+    auto* ra = new uint64_t[np << logN];
+    auto* rb = new uint64_t[np << logN];
+
+    ring.CRT(ra, cipher.ax, np);
+    ring.CRT(rb, cipher.bx, np);
+
+    ring.squareNTT(res.cx, ra, np, q);
+    ring.squareNTT(res.bx, rb, np, q);
+    ring.multDNTT(res.ax, ra, rb, np, q);
+    ring.addAndEqual(res.ax, res.ax, q);
+
+    delete[] ra;
+    delete[] rb;
+}
+
+void Scheme::squareRawInplace(Ciphertext &res){
+    squareRaw(res, res);
+}
+
+// inplace-safe
+void Scheme::relinearize(Ciphertext &res, const Ciphertext &cipher){
+    if(!cipher.cx){
+        if(&res != &cipher)
+            res.copy(cipher);
+        return;
+    }
+    res.copyParams(cipher);
+
+    Key* key = isSerialized ? SerializationUtils::readKey(serKeyMap.at(MULTIPLICATION)) : keyMap.at(MULTIPLICATION);
+
+    ZZ qQ = ring.qpows[cipher.logq + logQ];
+    ZZ q = ring.qpows[cipher.logq];
+    ZZ* as = new ZZ[N];
+    ZZ* bs = new ZZ[N];
+
+    long np = ceil((cipher.logq + logQQ + logN + 2)/(double)pbnd);
+    auto* raa = new uint64_t[np << logN];
+    ring.CRT(raa, cipher.cx, np);
+    ring.multDNTT(as, raa, key->rax, np, qQ);
+    ring.multDNTT(bs, raa, key->rbx, np, qQ);
+
+    ring.rightShiftAndEqual(as, logQ);
+    ring.rightShiftAndEqual(bs, logQ);
+
+    ring.add(res.bx, cipher.bx, bs, q);
+    ring.add(res.ax, cipher.ax, as, q);
+
+    delete[] as;
+    delete[] bs;
+    res.dropCX();
+}
+
+void Scheme::relinearizeInplace(Ciphertext &cipher) {
+    relinearize(cipher, cipher);
+}
+
 
 //-----------------------------------------
 
+void Scheme::scalingUp(Ciphertext &res, Ciphertext &cipher, long dlogp, bool modup) {
+    res.copyParams(cipher);
+    res.logp += dlogp;
+    if(modup)
+        res.logq += dlogp;
+    auto newq = ring.qpows[res.logq];
+    ring.leftShift(res.ax, cipher.ax, dlogp, newq);
+    ring.leftShift(res.bx, cipher.bx, dlogp, newq);
+    if(cipher.cx){
+        res.addCX();
+        ring.leftShift(res.cx, cipher.cx, dlogp, newq);
+    }
+    else
+        res.dropCX();
+}
+
+void Scheme::equalize(Ciphertext &cipher1, Ciphertext &cipher2) {
+    if(cipher1.logp != cipher2.logp){
+        throw std::invalid_argument("scaling factor mismatch");
+    }
+    if(cipher1.logq > cipher2.logq){
+        modDownToAndEqual(cipher1, cipher2.logq);
+    }
+    else if(cipher1.logq < cipher2.logq){
+        modDownToAndEqual(cipher2, cipher1.logq);
+    }
+}
+
 void Scheme::reScaleBy(Ciphertext& res, Ciphertext& cipher, long dlogq) {
 	res.copyParams(cipher);
 	ring.rightShift(res.ax, cipher.ax, dlogq);
 	ring.rightShift(res.bx, cipher.bx, dlogq);
 	res.logp -= dlogq;
 	res.logq -= dlogq;
+    if(cipher.cx){
+        res.addCX();
+        ring.rightShift(res.cx, cipher.cx, dlogq);
+    }
+    else
+        res.dropCX();
 }
 
 void Scheme::reScaleTo(Ciphertext& res, Ciphertext& cipher, long logq) {
@@ -876,6 +1088,8 @@ void Scheme::reScaleByAndEqual(Ciphertext& cipher, long dlogq) {
 	ring.rightShiftAndEqual(cipher.bx, dlogq);
 	cipher.logq -= dlogq;
 	cipher.logp -= dlogq;
+    if(cipher.cx)
+        ring.rightShiftAndEqual(cipher.cx, dlogq);
 }
 
 void Scheme::reScaleToAndEqual(Ciphertext& cipher, long logq) {
@@ -884,6 +1098,8 @@ void Scheme::reScaleToAndEqual(Ciphertext& cipher, long logq) {
 	ring.rightShiftAndEqual(cipher.bx, dlogq);
 	cipher.logq = logq;
 	cipher.logp -= dlogq;
+    if(cipher.cx)
+        ring.rightShiftAndEqual(cipher.cx, dlogq);
 }
 
 void Scheme::modDownBy(Ciphertext& res, Ciphertext& cipher, long dlogq) {
@@ -892,6 +1108,12 @@ void Scheme::modDownBy(Ciphertext& res, Ciphertext& cipher, long dlogq) {
 	ring.mod(res.ax, cipher.ax, q);
 	ring.mod(res.bx, cipher.bx, q);
 	res.logq -= dlogq;
+    if(cipher.cx){
+        res.addCX();
+        ring.mod(res.cx, cipher.cx, q);
+    }
+    else
+        res.dropCX();
 }
 
 void Scheme::modDownByAndEqual(Ciphertext& cipher, long dlogq) {
@@ -899,6 +1121,8 @@ void Scheme::modDownByAndEqual(Ciphertext& cipher, long dlogq) {
 	ring.modAndEqual(cipher.ax, q);
 	ring.modAndEqual(cipher.bx, q);
 	cipher.logq -= dlogq;
+    if(cipher.cx)
+        ring.modAndEqual(cipher.cx, q);
 }
 
 void Scheme::modDownTo(Ciphertext& res, Ciphertext& cipher, long logq) {
@@ -907,6 +1131,12 @@ void Scheme::modDownTo(Ciphertext& res, Ciphertext& cipher, long logq) {
 	ring.mod(res.ax, cipher.ax, q);
 	ring.mod(res.bx, cipher.bx, q);
 	res.logq = logq;
+    if(cipher.cx){
+        res.addCX();
+        ring.mod(res.cx, cipher.cx, q);
+    }
+    else
+        res.dropCX();
 }
 
 void Scheme::modDownToAndEqual(Ciphertext& cipher, long logq) {
@@ -914,6 +1144,8 @@ void Scheme::modDownToAndEqual(Ciphertext& cipher, long logq) {
 	cipher.logq = logq;
 	ring.modAndEqual(cipher.ax, q);
 	ring.modAndEqual(cipher.bx, q);
+    if(cipher.cx)
+        ring.modAndEqual(cipher.cx, q);
 }
 
 
diff --git a/HEAAN/src/Scheme.h b/HEAAN/src/Scheme.h
index 0c41b52..ba77592 100644
--- a/HEAAN/src/Scheme.h
+++ b/HEAAN/src/Scheme.h
@@ -118,51 +118,53 @@ public:
 	//   HOMOMORPHIC OPERATIONS
 	//----------------------------------------------------------------------------------
 
-	void negate(Ciphertext& res, Ciphertext& cipher);
+	void negate(Ciphertext& res, const Ciphertext &cipher);
 
 	void negateAndEqual(Ciphertext& cipher);
 
-	void add(Ciphertext& res, Ciphertext& cipher1, Ciphertext& cipher2);
+	void add(Ciphertext& res, const Ciphertext &cipher1, const Ciphertext &cipher2);
 
-	void addAndEqual(Ciphertext& cipher1, Ciphertext& cipher2);
+	void addAndEqual(Ciphertext& cipher1, const Ciphertext &cipher2);
 
-	void addConst(Ciphertext& res, Ciphertext& cipher, double cnst, long logp);
+	void addConst(Ciphertext& res, const Ciphertext &cipher, double cnst, long logp);
 
-	void addConst(Ciphertext& res, Ciphertext& cipher, NTL::RR& cnst, long logp);
+	void addConst(Ciphertext& res, const Ciphertext &cipher, const NTL::RR &cnst, long logp);
 
-	void addConst(Ciphertext& res, Ciphertext& cipher, std::complex<double> cnst, long logp);
+	void addConst(Ciphertext& res, const Ciphertext &cipher, std::complex<double> cnst, long logp);
 
 	void addConstAndEqual(Ciphertext& cipher, double cnst, long logp);
 
-	void addConstAndEqual(Ciphertext& cipher, NTL::RR& cnst, long logp);
+	void addConstAndEqual(Ciphertext& cipher, const NTL::RR &cnst, long logp);
 
 	void addConstAndEqual(Ciphertext& cipher, std::complex<double> cnst, long logp);
 
-	void sub(Ciphertext& res, Ciphertext& cipher1, Ciphertext& cipher2);
+	void sub(Ciphertext& res, const Ciphertext &cipher1, const Ciphertext &cipher2);
 
-	void subAndEqual(Ciphertext& cipher1, Ciphertext& cipher2);
+	void subAndEqual(Ciphertext& cipher1, const Ciphertext &cipher2);
 
-	void subAndEqual2(Ciphertext& cipher1, Ciphertext& cipher2);
+	void subAndEqual2(Ciphertext& cipher1, const Ciphertext &cipher2);
 
-	void imult(Ciphertext& res, Ciphertext& cipher);
+	void imult(Ciphertext& res, const Ciphertext &cipher);
 
-	void idiv(Ciphertext& res, Ciphertext& cipher);
+	void idiv(Ciphertext& res, const Ciphertext &cipher);
 
 	void imultAndEqual(Ciphertext& cipher);
 
 	void idivAndEqual(Ciphertext& cipher);
 
-	void mult(Ciphertext& res, Ciphertext& cipher1, Ciphertext& cipher2);
+	void mult(Ciphertext& res, const Ciphertext &cipher1, const Ciphertext &cipher2);
 
-	void multAndEqual(Ciphertext& cipher1, Ciphertext& cipher2);
+	void multAndEqual(Ciphertext& cipher1, const Ciphertext &cipher2);
 
-	void square(Ciphertext& res, Ciphertext& cipher);
+	void square(Ciphertext& res, const Ciphertext &cipher);
 
 	void squareAndEqual(Ciphertext& cipher);
 
-	void multByConst(Ciphertext& res, Ciphertext& cipher, double cnst, long logp);
+	void multByConst(Ciphertext& res, const Ciphertext &cipher, double cnst, long logp);
 
-	void multByConst(Ciphertext& res, Ciphertext& cipher, std::complex<double> cnst, long logp);
+    void multByConst(Ciphertext& res, const Ciphertext &cipher, const NTL::RR& cnst, long logp);
+
+	void multByConst(Ciphertext& res, const Ciphertext &cipher, std::complex<double> cnst, long logp);
 
 	void multByConstVec(Ciphertext& res, Ciphertext& cipher, std::complex<double>* cnstVec, long logp);
 
@@ -170,7 +172,7 @@ public:
 
 	void multByConstAndEqual(Ciphertext& cipher, double cnst, long logp);
 
-	void multByConstAndEqual(Ciphertext& cipher, NTL::RR& cnst, long logp);
+	void multByConstAndEqual(Ciphertext& cipher, const NTL::RR& cnst, long logp);
 
 	void multByConstAndEqual(Ciphertext& cipher, std::complex<double> cnst, long logp);
 
@@ -196,11 +198,32 @@ public:
 
 	void divByPo2AndEqual(Ciphertext& cipher, long bits);
 
+    // ******************* raw multiplications **********************
+
+    void matchScalingFactor(Ciphertext& dst1, Ciphertext& dst2, const Ciphertext& src1, const Ciphertext& src2, long logp);
+
+    void multRaw(Ciphertext& res, const Ciphertext &cipher1, const Ciphertext &cipher2);
+
+    void multRawInplace(Ciphertext& cipher1, const Ciphertext &cipher2);
+
+    void squareRaw(Ciphertext& res, const Ciphertext &cipher);
+
+    void squareRawInplace(Ciphertext &res);
+
+    void relinearize(Ciphertext &res, const Ciphertext &cipher);
+
+    void relinearizeInplace(Ciphertext &cipher);
+
 
 	//----------------------------------------------------------------------------------
 	//   RESCALING
 	//----------------------------------------------------------------------------------
 
+    // increase the scaling factor by `dlogp` bits, optionally increase the modulus by modup bits.
+    void scalingUp(Ciphertext& res, Ciphertext& cipher, long dlogp, bool modup=false);
+
+    // bring both ciphertexts to the same modulus (their scaling factor must be the same)
+    void equalize(Ciphertext &cipher1, Ciphertext &cipher2);
 
 	void reScaleBy(Ciphertext& res, Ciphertext& cipher, long dlogq);
 
